File: sequence-guessing-game/src/SequenceGuessingGame.js
import React, { useState } from 'react';


const generateRandomSequences = () => {
  const sequences = [];
  const usedSequences = new Set();
  const sequenceToString = (seq) => seq.join(',');
  const numCases = 10;
  const caseTypes = Array.from({ length: numCases }, (_, i) => i);

  // Shuffle the case types
  for (let i = caseTypes.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [caseTypes[i], caseTypes[j]] = [caseTypes[j], caseTypes[i]];
  }

  for (const type of caseTypes) {
    
    let sequence = [];
    let description = '';
    let formula = '';
    let nextNumber = 0; 
    

    switch (type) {
      case 0:
        {
          let divisor = Math.floor(Math.random() * 3) + 2;
          let addend = Math.floor(Math.random() * 6);
          let start = (Math.floor(Math.random() * 20) + 1) * divisor;
      
          sequence = Array.from({ length: 5 }, (_, idx) => {
            const n = start + idx * divisor;
            return (n / divisor) + addend;
          });
          nextNumber = (start + 5 * divisor) / divisor + addend;
          description = `Starting at ${start}, each number divided by ${divisor} plus ${addend}`;
          formula = `x[n] = ((${start} + n×${divisor}) ÷ ${divisor}) + ${addend}`;
        }
        break;
      case 1:
        {
        const start = Math.floor(Math.random() * 201) - 100;
        const ratio = Math.floor(Math.random() * 3) + 2;
        sequence = Array.from({ length: 5 }, (_, idx) => start * Math.pow(ratio, idx));
        nextNumber = sequence[sequence.length - 1] * ratio;  // Calculate next
        description = `Geometric sequence with ratio of ${ratio}`;
        formula = `x[n] = x[n-1] × ${ratio}`;
        }
        break;
      case 2:
        {
          const start = Math.floor(Math.random() * 201) - 100;
          const squareConst = Math.floor(Math.random() * 5) + 1;
          sequence = Array.from({ length: 5 }, (_, idx) => Math.pow(idx + 1, 2) + squareConst);
          nextNumber = Math.pow(6, 2) + squareConst;  // Calculate next
          description = `Square numbers plus ${squareConst}`;
          formula = `x[n] = n² + ${squareConst}`;
        }
        break;
      case 3:
        {
          const start = Math.floor(Math.random() * 201) - 100;
          const factorial = (n) => (n === 0 || n === 1 ? 1 : n * factorial(n - 1));
          const isAddFactorial = Math.random() < 0.5; // 50% chance for addition vs subtraction
          const factorialOffset = Math.floor(Math.random() * 5) + 1;
          
          sequence = Array.from({ length: 5 }, (_, idx) => {
            const fact = factorial(idx + 1);
            return isAddFactorial ? fact + factorialOffset : fact - factorialOffset;
          });
          nextNumber = isAddFactorial ? factorial(6) + factorialOffset : factorial(6) - factorialOffset;
          description = `Factorial sequence ${isAddFactorial ? 'plus' : 'minus'} ${factorialOffset}`;
          formula = `x[n] = n! ${isAddFactorial ? '+' : '-'} ${factorialOffset}`;
        }
        break;
      case 4:
        {
          const start = Math.floor(Math.random() * 201) - 100;
          const addNum = Math.floor(Math.random() * 5) + 1;
          const isAdd = Math.random() < 0.5;
          // 30% chance to allow negative numbers
          const allowNegStart = Math.random() < 0.3;
          const minStartValue = allowNegStart ? -30 : (isAdd ? 1 : addNum + 1);
          start = Math.floor(Math.random() * 50) + minStartValue;
          sequence = Array.from({ length: 5 }, (_, idx) => start + (idx * 3) + (isAdd ? addNum : -addNum));
          nextNumber = start + (5 * 3) + (isAdd ? addNum : -addNum);
          description = `Multiply by 3 ${isAdd ? 'and add' : 'and subtract'} ${addNum}`;
          formula = `x[n] = ${start} + 3n ${isAdd ? '+' : '-'} ${addNum}`;
        }
        break;
      case 5:
        {
          const squareAdd = Math.floor(Math.random() * 5) + 1;
          const isPlus = Math.random() < 0.5; 
          const start = Math.floor(Math.random() * 201) - 100; 

          const sequence = Array.from({ length: 5 }, (_, idx) => Math.pow(start + idx + 1, 2) + (isPlus ? squareAdd : -squareAdd));
          const nextNumber = Math.pow(start + 6, 2) + (isPlus ? squareAdd : -squareAdd);
    
          const description = `Square number ${isPlus ? 'plus' : 'minus'} ${squareAdd}`;
          const formula = `x[n] = (${start} + n)² ${isPlus ? '+' : '-'} ${squareAdd}`;
        }
        break;
      case 6:  
        {
          let start = Math.floor(Math.random() * 201) - 100;
          let cubeConst = Math.floor(Math.random() * 5) + 1;
          let sequence = Array.from({ length: 5 }, (_, idx) => Math.pow(idx + 1, 3) + cubeConst);
          let nextNumber = Math.pow(6, 3) + cubeConst;  // Calculate next
          let description = `Cube numbers plus ${cubeConst}`;
          let formula = `x[n] = n³ + ${cubeConst}`;
        }
        break;
      case 7:  // Add new case for square root sequence
        {
          const offset = Math.floor(Math.random() * 5); // Random offset 0-4
          sequence = Array.from({ length: 5 }, (_, idx) => {
            const perfectSquare = Math.pow(idx + offset + 1, 2); // Generate perfect square
            return Math.sqrt(perfectSquare);
          });
          nextNumber = Math.sqrt(Math.pow(5 + offset + 1, 2)); // Calculate next
          description = `Square root of perfect squares starting from ${Math.pow(offset + 1, 2)}`;
          formula = `x[n] = √(${offset > 0 ? '(n+' + offset + ')' : 'n'}²)`;
        }
        break;
      case 8: 
        {
        // Generate random start between -1000 and 1000
        const fibStart = Math.floor(Math.random() * 2001) - 1000;        
        const fibSecond = fibStart + (Math.floor(Math.random() * 21) - 10);

        
        sequence = [fibStart, fibSecond];
        // Generate next 3 numbers in Fibonacci-style (each number is sum of previous two)
        for (let i = 2; i < 5; i++) {
          sequence.push(sequence[i-1] + sequence[i-2]);
        }
        nextNumber = sequence[sequence.length - 1] + sequence[sequence.length - 2];
        description = `Fibonacci-style sequence starting at ${fibStart}`;
        formula = `x[n] = x[n-1] + x[n-2]`;
        }
        break;
      case 9: // New case for a + [(n + 2) * 3]
        {
          const startingNumber = Math.floor(Math.random() * 201) - 100; // Random number between -100 and 100
        sequence = [startingNumber];
        for (let i = 1; i < 5; i++) {
          sequence.push(sequence[i - 1] + ((i + 2) * 3));
        }
        nextNumber = sequence[sequence.length - 1] + ((sequence.length + 2) * 3);
        description = 'Sequence with a starting number and adding (n + 2) * 3 to the previous number';
        formula = `x[n] = x[n-1] + (n + 2) * 3`;
        }
        break;
      default:
        sequence = Array.from({ length: 5 }, (_, idx) => idx + 1);
        nextNumber = 6;
        description = 'Simple counting sequence';
        formula = 'x[n] = n';
        break;
    }

    const sequenceString = sequenceToString(sequence);
    if (!usedSequences.has(sequenceString)) {
      usedSequences.add(sequenceString);
      sequences.push({ 
        sequence, 
        type, 
        description, 
        formula,
        nextNumber 
      });
    }
  }

  return sequences;
};

const SequenceGuessingGame = () => {
  const [sequences] = useState(generateRandomSequences());
  const [guesses, setGuesses] = useState(Array(10).fill(''));
  const [results, setResults] = useState(Array(10).fill(null));
  const [attempts, setAttempts] = useState(Array(10).fill(0)); 

  const handleGuessChange = (index, value) => {
    const newGuesses = [...guesses];
    newGuesses[index] = value;
    setGuesses(newGuesses);
  };

  const evaluateGuess = (input) => {
    // Remove all whitespace
    const sanitizedInput = input.replace(/\s+/g, '');
    
    // Check if input contains only valid mathematical expressions
    const validPattern = /^[0-9+\-*/().]+$/;
    
    if (!validPattern.test(sanitizedInput)) {
      return parseFloat(input); // Return original input if it's not a valid expression
    }
  
    try {
      // Evaluate the expression
      const result = eval(sanitizedInput);
      return typeof result === 'number' && !isNaN(result) ? result : parseFloat(input);
    } catch (error) {
      return parseFloat(input); // Return original input if evaluation fails
    }
  };
  
  const checkGuess = (index) => {
    const userInput = guesses[index];
    const userGuess = evaluateGuess(userInput);
    const expectedNext = sequences[index].nextNumber;  
  
    
    console.log('Checking sequence:', {
      sequence: sequences[index].sequence,
      type: sequences[index].type,
      description: sequences[index].description,
      formula: sequences[index].formula,
      expectedNext,
      userGuess,
      attempts: attempts[index] + 1
    });
  
    const newResults = [...results];
    const newAttempts = [...attempts];
  
    newAttempts[index]++;
    setAttempts(newAttempts);
  
    newResults[index] = Math.abs(userGuess - expectedNext) < 0.1;
    setResults(newResults);
  };
  

  
  return (
    <div style={{ padding: '20px', fontFamily: 'Arial, sans-serif', backgroundColor: '#000', color: '#fff', minHeight: '100vh' }}>
      <h1 style={{ textAlign: 'center', color: '#00ff00' }}>Sequence Guessing Game</h1>
      <p style={{ textAlign: 'center' }}>Below are 10 sequences. Can you guess the next number?</p>
      <ul style={{ listStyleType: 'none', padding: 0 }}>
        {sequences.map((seqData, index) => (
          <li key={index} style={{ marginBottom: '15px', padding: '10px', border: '1px solid #00ff00', borderRadius: '5px' }}>
            {seqData.sequence.join(', ')}
            <br />
            <div style={{ display: 'flex', gap: '10px', marginTop: '5px' }}>
              <input
                type="text"
                placeholder="Your guess for the next number"
                value={guesses[index]}
                onChange={(e) => handleGuessChange(index, e.target.value)}
                style={{ padding: '5px', flexGrow: 1, boxSizing: 'border-box' }}
              />
              <button 
                onClick={() => checkGuess(index)}
                style={{ 
                  padding: '5px 10px', 
                  backgroundColor: '#00ff00', 
                  color: '#000', 
                  border: 'none', 
                  borderRadius: '3px',
                  cursor: 'pointer'
                }}
              >
                Check
              </button>
            </div>
            {results[index] !== null && (
              <div style={{ 
                marginTop: '5px', 
                color: results[index] ? '#00ff00' : '#ff0000' 
              }}>
                {results[index] ?  (
                    <>
                      Correct! <br/>
                      <span style={{ fontSize: '0.9em', opacity: '0.8' }}>
                        Formula: {seqData.formula}
                      </span>
                    </>
                  ) : (
                  <>
                    Try again! <br/>
                    <span style={{ fontSize: '0.9em', opacity: '0.8' }}>
                    Hint: This is a {seqData.description}
                      {attempts[index] >= 3 && (
                        <>
                          <br/>
                          Formula: {seqData.formula}
                        </>
                      )}
                    </span>
                  </>
                )}
              </div>
            )}
          </li>
        ))}
      </ul>
      <p style={{ textAlign: 'center' }}>Refresh the page to get new sequences!</p>
    </div>
  );
};

export default SequenceGuessingGame;

File: sequence-guessing-py/SequenceGuessingGame.py
import random
import math
import streamlit as st

def generate_divided_and_added_sequence():
    divisor = random.randint(2, 4)
    start = (random.randint(1, 20) + 1) * divisor
    addend = random.randint(0, 5)

    sequence = [(start + idx * divisor) // divisor + addend for idx in range(5)]
    next_number = (start + 5 * divisor) // divisor + addend
    description = f"Starting at {start}, each number divided by {divisor} plus {addend}"
    formula = f"x[n] = ((({start} + n×{divisor}) ÷ {divisor}) + {addend})"

    return sequence, next_number, description, formula

def generate_geometric_sequence():
    ratio = random.randint(2, 4)
    start = random.randint(-100, 100)
    sequence = [start * ratio ** idx for idx in range(5)]
    next_number = sequence[-1] * ratio
    description = f"Geometric sequence with ratio of {ratio}"
    formula = f"x[n] = {start} × {ratio}^n"

    return sequence, next_number, description, formula

def generate_square_sequence():
    square_const = random.randint(1, 5)
    sequence = [(idx + 1) ** 2 + square_const for idx in range(5)]
    next_number = 6 ** 2 + square_const
    description = f"Square numbers plus {square_const}"
    formula = f"x[n] = n² + {square_const}"

    return sequence, next_number, description, formula

def generate_factorial_sequence():
    factorial_offset = random.randint(1, 5)
    is_add_factorial = random.choice([True, False])
    sequence = [math.factorial(idx + 1) + factorial_offset if is_add_factorial else math.factorial(idx + 1) - factorial_offset for idx in range(5)]
    next_number = math.factorial(6) + factorial_offset if is_add_factorial else math.factorial(6) - factorial_offset
    description = f"Factorial sequence {'plus' if is_add_factorial else 'minus'} {factorial_offset}"
    formula = f"x[n] = n! {'+ ' if is_add_factorial else '- '}{factorial_offset}"

    return sequence, next_number, description, formula

def generate_multiply_and_add_sequence():
    add_num = random.randint(1, 5)
    is_add = random.choice([True, False])
    allow_neg_start = random.random() < 0.3
    min_start_value = 1 if is_add else add_num + 1
    start = random.randint(min_start_value if not allow_neg_start else -30, 50)
    sequence = [start + (idx * 3) + (add_num if is_add else -add_num) for idx in range(5)]
    next_number = start + (5 * 3) + (add_num if is_add else -add_num)
    description = f"Multiply by 3 {'and add' if is_add else 'and subtract'} {add_num}"
    formula = f"x[n] = {start} + 3n {'+ ' if is_add else '- '}{add_num}"

    return sequence, next_number, description, formula

def generate_square_plus_minus_sequence():
    square_add = random.randint(1, 5)
    is_plus = random.choice([True, False])
    start = random.randint(-100, 100)
    sequence = [pow(start + idx + 1, 2) + (square_add if is_plus else -square_add) for idx in range(5)]
    next_number = pow(start + 6, 2) + (square_add if is_plus else -square_add)
    description = f"Square number {'plus' if is_plus else 'minus'} {square_add}"
    formula = f"x[n] = ({start} + n)² {'+ ' if is_plus else '- '}{square_add}"

    return sequence, next_number, description, formula

def generate_cube_sequence():
    cube_const = random.randint(1, 5)
    sequence = [(idx + 1) ** 3 + cube_const for idx in range(5)]
    next_number = 6 ** 3 + cube_const
    description = f"Cube numbers plus {cube_const}"
    formula = f"x[n] = n³ + {cube_const}"

    return sequence, next_number, description, formula

def generate_square_root_sequence():
    offset = random.randint(0, 4)
    sequence = [int(math.sqrt(pow(idx + offset + 1, 2))) for idx in range(5)]
    next_number = int(math.sqrt(pow(5 + offset + 1, 2)))
    description = f"Square root of perfect squares starting from {pow(offset + 1, 2)}"
    formula = f"x[n] = √({'(n+' + str(offset) + ')' if offset > 0 else 'n'}²)"

    return sequence, next_number, description, formula

def generate_fibonacci_sequence():
    fib_sequence = [1, 1]
    for _ in range(21):
        fib_sequence.append(fib_sequence[-1] + fib_sequence[-2])

    # Pick a random starting index between 0 and 1000
    start_index = random.randint(0, 19) 

    # Get the sequence of 5 consecutive Fibonacci numbers starting from the random index
    sequence = fib_sequence[start_index:start_index+5]
    next_number = fib_sequence[start_index+5]
    description = f"Fibonacci sequence starting from the {start_index}th number"
    formula = "x[n] = x[n-1] + x[n-2]"

    return sequence, next_number, description, formula

def generate_starting_number_sequence():
    starting_number = random.randint(-100, 100)
    sequence = [starting_number]
    for idx in range(1, 5):
        sequence.append(sequence[-1] + (idx + 2) * 3)
    next_number = sequence[-1] + (len(sequence) + 2) * 3
    description = "Sequence with a starting number and adding (n + 2) * 3 to the previous number"
    formula = "x[n] = x[n-1] + (n + 2) * 3"

    return sequence, next_number, description, formula

def generate_random_sequences():
    sequence_generators = [
        generate_divided_and_added_sequence,
        generate_geometric_sequence,
        generate_square_sequence,
        generate_factorial_sequence,
        generate_multiply_and_add_sequence,
        generate_square_plus_minus_sequence,
        generate_cube_sequence,
        generate_square_root_sequence,
        generate_fibonacci_sequence,
        generate_starting_number_sequence
    ]

    sequences = []
    used_sequences = set()
    sequence_to_string = lambda seq: ','.join(map(str, seq))

    case_types = list(range(10))
    random.shuffle(case_types)

    for case_type in case_types:
        sequence, next_number, description, formula = sequence_generators[case_type]()
        sequence_string = sequence_to_string(sequence)

        # {{ Check if the sequence is a straight sequence }}
        if len(set(sequence)) == len(sequence) and sequence == list(range(min(sequence), max(sequence)+1)):
            continue  # Skip this sequence and generate a new one
        # {{ /Check if the sequence is a straight sequence }}
        
        if sequence_string not in used_sequences:
            used_sequences.add(sequence_string)
            sequences.append({
                'sequence': sequence,
                'type': case_type,
                'description': description,
                'formula': formula,
                'next_number': next_number
            })

    return sequences

def main():
    st.set_page_config(page_title="Sequence Guessing Game", layout="wide")
    st.title("Sequence Guessing Game")
    st.write("Below are 10 sequences. Can you guess the next number?")

    if "sequences" not in st.session_state:
        st.session_state.sequences = generate_random_sequences()
        st.session_state.user_guesses = [None] * len(st.session_state.sequences)
        st.session_state.formula_shown = [False] * len(st.session_state.sequences)

    sequences = st.session_state.sequences


    for idx, seq_data in enumerate(sequences):
        with st.expander(f"Sequence {idx + 1}"):
            st.write(", ".join(map(str, seq_data['sequence'])))
            user_guess = st.text_input(f"Your guess for the next number", key=f"guess_{idx}", value=str(st.session_state.user_guesses[idx]) if st.session_state.user_guesses[idx] is not None else "")
            attempt_count = sum(1 for prev_guess in st.session_state.user_guesses[:idx+1] if prev_guess is not None)
            if st.button(f"Check Sequence {idx + 1}", key=f"check_{idx}"):
                if not user_guess:
                    st.warning("Please enter a value to check the sequence.")
                else:
                    try:
                        user_guess = float(user_guess)
                        st.session_state.user_guesses[idx] = user_guess
                        if abs(user_guess - seq_data['next_number']) < 0.1:
                            st.success(f"Correct! Formula: {seq_data['formula']}")
                            st.session_state.formula_shown[idx] = True
                        else:
                            if attempt_count < 2:
                                st.error(f"Try again! Hint: This is a {seq_data['description']}")
                            else:
                                st.error(f"Try again! Formula: {seq_data['formula']}")
                                st.session_state.formula_shown[idx] = True
                    except ValueError:
                        st.warning("Please enter a valid number.")


if __name__ == "__main__":
    main()